musicbot/billboard/Billboard.py

```diff
@@ -0,0 +1,174 @@
+        self._initializeWorkingDirectories()
+        self._loadBillboardFile()
+
+    def _throwHelpfulErrorIfWorkingDirectoryDoesNotExist(self):
+        if not Path(self._base_working_directory).exists():
+            raise HelpfulError("Unable to proceed. Cannot find directory: {} ".format(str(self._base_working_directory)))
+        
+    def _initializeWorkingDirectories(self):
+        self._working_directory = self._base_working_directory + "billboard\\" + self._billboard_name + "\\"
+        self._song_working_directory = self._working_directory + "songs\\"
+
+        log.debug("_initializeWorkingDirectories called. working_directory: {}, song_working_dictionary: {}".format(
+            self._working_directory,
+            self._song_working_directory
+        ))
+        
+        Path(self._working_directory).mkdir(exist_ok=True)
+        Path(self._song_working_directory).mkdir(exist_ok=True)
+        
+    def _loadBillboardFile(self):
+        billboard_json_file_path = self._working_directory + "info.json"
+        self._billboard_content = {
+            "date_last_calculated" : "1970-01-01",
+            "song_ids_ordered_by_most_to_least_queued" : []
+        }
+        log.debug("_loadBillboardFile called. billboard_json_file_path: {}".format(billboard_json_file_path))
+        self._json_file = JsonFile(self._working_directory + "info.json", default_content = self._billboard_content)
+        json_content = self._json_file.json()
+
+        for key, value in self._billboard_content.items():
+            self._billboard_content[key] = value if key not in json_content else json_content[key]
+
+
+    def _calculateBillboardTopSongs(self):
+        log.debug("_calculateBillboardTopSongs called.")
+        self._top_songs_cache = []
+        song_manager = SongManager()
+
+        for video_id in next(os.walk(self._song_working_directory))[1]:
+            log.debug("video_id: {}".format(video_id))
+            song_manager.manage(
+                Song({
+                    "video_id": video_id,
+                    "title" : "null"
+                }, 
+                self._song_working_directory))
+
+        self._billboard_content["song_ids_ordered_by_most_to_least_queued"] = []
+        top_queued_songs = song_manager.getTopQueuedSongs(self._number_of_songs_to_display)
+        for song in top_queued_songs:
+            self._top_songs_cache.append(song)
+            
+            song.incrementNumberOfWeeksInChart()
+            song.updatePositionOnChart(len(self._top_songs_cache))
+            song.save()
+            log.debug("song: {}".format(str(song)))
+            
+            self._billboard_content["song_ids_ordered_by_most_to_least_queued"].append(song.getVideoId()) 
+            
+    def _loadTopSongsToCacheIfNeeded(self):
+        log.debug("_loadTopSongsToCacheIfNeeded called.")
+    
+        if len(self._top_songs_cache) == 0:
+            song_manager = SongManager()
```

Comment:
When is the instance of SongManager deleted? I think it should also become a Generator if its scope is only within select events/functions.

Reply:
a Generator is supposed to save on memory by only creating objects on-the-fly. 

in this case, a Generator will not work because all the songs need to be loaded to compare which ones should be displayed first